From 44ae9e33525207cc875296d8baf33cef23f4a115 Mon Sep 17 00:00:00 2001
From: Daniel Rodgers-Pryor <djrodgerspryor@gmail.com>
Date: Sun, 7 Sep 2014 15:33:59 +1000
Subject: [PATCH] Allow for 2 periodic axes in 3d

Added subclasses of _BasePeriodicGrid3D to describe two dimensions of
periodicity.

re #671
---
 fipy/meshes/periodicGrid3D.py | 25 ++++++++++++++++++++++++-
 1 file changed, 24 insertions(+), 1 deletion(-)

diff --git a/fipy/meshes/periodicGrid3D.py b/fipy/meshes/periodicGrid3D.py
index e478d09..de9ef6f 100755
--- a/fipy/meshes/periodicGrid3D.py
+++ b/fipy/meshes/periodicGrid3D.py
@@ -41,7 +41,9 @@ from fipy.tools import numerix
 from fipy.tools import parallelComm
 from fipy.meshes.nonUniformGrid3D import NonUniformGrid3D
 
-__all__ = ["PeriodicGrid3D", "PeriodicGrid3DLeftRight", "PeriodicGrid3DTopBottom", "PeriodicGrid3DFrontBack"]
+__all__ = ["PeriodicGrid3D", "PeriodicGrid3DLeftRight", "PeriodicGrid3DTopBottom",
+           "PeriodicGrid3DFrontBack", "PeriodicGrid3DLeftRightTopBottom",
+           "PeriodicGrid3DLeftRightFrontBack", "PeriodicGrid3DTopBottomFrontBack"]
 
 class _BasePeriodicGrid3D(NonUniformGrid3D):
     def __init__(self, dx=1., dy=1., dz=1., nx=None, ny=None, nz=None, overlap=2, communicator=parallelComm, *args, **kwargs):
@@ -174,11 +176,32 @@ class PeriodicGrid3DLeftRight(_BasePeriodicGrid3D):
         self._connectFaces(numerix.nonzero(self.facesLeft),
                            numerix.nonzero(self.facesRight))
 
+class PeriodicGrid3DLeftRightTopBottom(_BasePeriodicGrid3D):
+    def _makePeriodic(self):
+        self._connectFaces(numerix.nonzero(self.facesLeft),
+                           numerix.nonzero(self.facesRight))
+        self._connectFaces(numerix.nonzero(self.facesBottom),
+                           numerix.nonzero(self.facesTop))
+
+class PeriodicGrid3DLeftRightFrontBack(_BasePeriodicGrid3D):
+    def _makePeriodic(self):
+        self._connectFaces(numerix.nonzero(self.facesLeft),
+                           numerix.nonzero(self.facesRight))
+        self._connectFaces(numerix.nonzero(self.facesFront),
+                           numerix.nonzero(self.facesBack))
+
 class PeriodicGrid3DTopBottom(_BasePeriodicGrid3D):
     def _makePeriodic(self):
         self._connectFaces(numerix.nonzero(self.facesBottom),
                            numerix.nonzero(self.facesTop))
 
+class PeriodicGrid3DTopBottomFrontBack(_BasePeriodicGrid3D):
+    def _makePeriodic(self):
+        self._connectFaces(numerix.nonzero(self.facesBottom),
+                           numerix.nonzero(self.facesTop))
+        self._connectFaces(numerix.nonzero(self.facesFront),
+                           numerix.nonzero(self.facesBack))
+
 class PeriodicGrid3DFrontBack(_BasePeriodicGrid3D):
     def _makePeriodic(self):
         self._connectFaces(numerix.nonzero(self.facesFront),
-- 
1.8.3.msysgit.0

